Wikipedia/Code/WebViewController.m

```diff
@@ -89,146 +90,164 @@ - (instancetype)initWithSession:(SessionSingleton*)aSession {
 #pragma mark - WKScriptMessageHandler
 
 - (void)userContentController:(WKUserContentController*)userContentController didReceiveScriptMessage:(WKScriptMessage*)message {
-    /**
-     *  Warning!!
-     *  All messages from JS can contain NSNulls.
-     *  All elements must be checked for existence of NSNull before being passed on to any other methods.
-     * Additionally body is untyped, so its class must be
-     * checked before proceeding.
-     */
-
-    if ([message.body isKindOfClass:[NSNull class]]) {
+
+    WMFWKScriptMessageType messageType = [WKScriptMessage wmf_typeForMessageName:message.name];
+    if (![message.body isKindOfClass:[WKScriptMessage wmf_expectedMessageBodyClassForType:messageType]]) {
+        NSAssert(NO, @"Unexpected script message body kind of class!");
         return;
     }
 
-    if ([message.body isKindOfClass:[NSDictionary class]]) {
-        
-        NSDictionary* body = [message.body wmf_dictionaryByRemovingNullObjects];
-        
-        if ([message.name isEqualToString:@"peek"]) {
-            NSDictionary* peekElementDict = body[@"peekElement"];
-            
-            if(![peekElementDict isKindOfClass:[NSDictionary class]]){
-                return;
+    switch (messageType) {
+        case WMFWKScriptMessagePeek:
+            [self handlePeekScriptMessage:message.body];
+            break;
+        case WMFWKScriptMessageConsoleMessage:
+            [self handleMessageConsoleScriptMessage:message.body];
+            break;
+        case WMFWKScriptMessageClickLink:
+            [self handleClickLinkScriptMessage:message.body];
+            break;
+        case WMFWKScriptMessageClickImage:
+            [self handleClickImageScriptMessage:message.body];
+            break;
+        case WMFWKScriptMessageClickReference:
+            [self handleClickReferenceScriptMessage:message.body];
+            break;
+        case WMFWKScriptMessageClickEdit:
+            [self handleClickEditScriptMessage:message.body];
+            break;
+        case WMFWKScriptMessageNonAnchorTouchEndedWithoutDragging:
+            [self handleNonAnchorTouchEndedWithoutDraggingScriptMessage];
+            break;
+        case WMFWKScriptMessageLateJavascriptTransform:
+            [self handleLateJavascriptTransformScriptMessage:message.body];
+            break;
+        case WMFWKScriptMessageArticleState:
+            [self handleArticleStateScriptMessage:message.body];
+            break;
+        case WMFWKScriptMessageUnknown:
+            NSAssert(NO, @"Unhandled script message type!");
+            break;
+    }
+}
+
+- (void)handlePeekScriptMessage:(NSDictionary*)messageDict {
+    if(messageDict.count == 0){
```

Comment:
Minor suggestion - check for the elements you want (tagName, src, href) rather than checking messageCount == 0 - this way the behavior is safe even if some objects are missing or the wrong keys are passed